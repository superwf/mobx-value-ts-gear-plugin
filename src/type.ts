export type Option = {
  /**
   * output file path
   * @example "src/request/req.ts"
   * */
  output: string
  /**
   * how to generate your mobx-value request var
   * the param is requestFunctionName generate by ts-gear
   * */
  generateRequestVarName?: (requestFunctionName: string) => string

  /**
   * where the output and service dir is based
   * @default "src"
   * */
  baseDir?: string
}

export type CollectRequestResult = {
  requestFunctionNames: string[]
  returnTypeNames: string[]
  defaultValues: string[]
}
