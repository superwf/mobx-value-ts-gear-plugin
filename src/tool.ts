import path from 'path'

export const resolveRoot = (...paths: string[]) => path.resolve(process.cwd(), ...paths)
