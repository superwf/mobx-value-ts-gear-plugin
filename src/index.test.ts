import fs from 'fs'
import path from 'path'
import temp from 'temp'
import { upperFirst } from 'lodash'
import type { Project } from 'ts-gear'
import type { Option } from './type'
import { resolveRoot } from './tool'
import { mobxValueTsGearPlugin } from './index'

test('use beforeWriteTs hook', async () => {
  expect(true).toBe(true)
  const tempDir = await temp.mkdir('tempForTest')
  const output = path.join(tempDir, 'output.ts')
  const option: Option = {
    output,
  }

  const requestFileContent = fs.readFileSync(resolveRoot('fixture/petv3/request.ts'), { encoding: 'utf8' })

  const project: Project = {
    name: 'petv3',
    dest: 'service',
    source: 'https://petstore3.swagger.io/api/v3/openapi.json',
    importRequesterStatement: 'import { requester } from "../../requester"',
    // simplifyRequestOption: true,
    shouldGenerateMock: true,
    EOL: '\n',
    withBasePath: true,
    requestOptionUnionType: 'RequestInit',
  }

  await mobxValueTsGearPlugin(option).beforeWriteTs!({
    requestFileContent,
    requestFile: '',
    definitionFile: '',
    definitionFileContent: '',
    mockRequestFile: '',
    mockRequestFileContent: '',
    indexFile: '',
    indexFileContent: '',
    project,
  })
  const result = fs.readFileSync(output, { encoding: 'utf8' })
  expect(result).toMatchSnapshot()
  temp.cleanupSync()
})

test('transformRequestFunctionName', async () => {
  expect(true).toBe(true)
  const tempDir = await temp.mkdir('tempForTest')
  const output = path.join(tempDir, 'output.ts')
  const option: Option = {
    output,
    transformRequestFunctionName(reqFuncName) {
      return `mobx${upperFirst(reqFuncName)}`
    },
  }

  const requestFileContent = fs.readFileSync(resolveRoot('fixture/petv3/request.ts'), { encoding: 'utf8' })

  const project: Project = {
    name: 'petv3',
    dest: 'service',
    source: 'https://petstore3.swagger.io/api/v3/openapi.json',
    importRequesterStatement: 'import { requester } from "../../requester"',
    // simplifyRequestOption: true,
    shouldGenerateMock: true,
    EOL: '\n',
    withBasePath: true,
    requestOptionUnionType: 'RequestInit',
  }

  await mobxValueTsGearPlugin(option).beforeWriteTs!({
    requestFileContent,
    requestFile: '',
    definitionFile: '',
    definitionFileContent: '',
    mockRequestFile: '',
    mockRequestFileContent: '',
    indexFile: '',
    indexFileContent: '',
    project,
  })
  const result = fs.readFileSync(output, { encoding: 'utf8' })
  // console.log(result)
  expect(result).toMatchSnapshot()
  temp.cleanupSync()
})
