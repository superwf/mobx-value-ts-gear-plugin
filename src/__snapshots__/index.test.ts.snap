// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`transformRequestFunctionName 1`] = `
"import { request } from \\"mobx-value\\";
import {
  putPet,
  postPet,
  getPetFindByStatus,
  getPetFindByTags,
  getPetPetId,
  postPetPetId,
  deletePetPetId,
  postPetPetIdUploadImage,
  getStoreInventory,
  postStoreOrder,
  getStoreOrderOrderId,
  deleteStoreOrderOrderId,
  postUser,
  postUserCreateWithList,
  getUserLogin,
  getUserLogout,
  getUserUsername,
  putUserUsername,
  deleteUserUsername,
  Pet,
  ApiResponse,
  Order,
  User,
} from \\"src/service/petv3\\";

/** auto generated by mobx-value-ts-gear-plugin */
export const mobxPutPet = request({
  request: putPet,
  value: {} as Pet,
});
/** auto generated by mobx-value-ts-gear-plugin */
export const mobxPostPet = request({
  request: postPet,
  value: {} as Pet,
});
/** auto generated by mobx-value-ts-gear-plugin */
export const mobxGetPetFindByStatus = request({
  request: getPetFindByStatus,
  value: {} as Pet[],
});
/** auto generated by mobx-value-ts-gear-plugin */
export const mobxGetPetFindByTags = request({
  request: getPetFindByTags,
  value: {} as Pet[],
});
/** auto generated by mobx-value-ts-gear-plugin */
export const mobxGetPetPetId = request({
  request: getPetPetId,
  value: {} as Pet,
});
/** auto generated by mobx-value-ts-gear-plugin */
export const mobxPostPetPetId = request({
  request: postPetPetId,
  value: {} as any,
});
/** auto generated by mobx-value-ts-gear-plugin */
export const mobxDeletePetPetId = request({
  request: deletePetPetId,
  value: {} as any,
});
/** auto generated by mobx-value-ts-gear-plugin */
export const mobxPostPetPetIdUploadImage = request({
  request: postPetPetIdUploadImage,
  value: {} as ApiResponse,
});
/** auto generated by mobx-value-ts-gear-plugin */
export const mobxGetStoreInventory = request({
  request: getStoreInventory,
  value: {} as { [propertyName: string]: number },
});
/** auto generated by mobx-value-ts-gear-plugin */
export const mobxPostStoreOrder = request({
  request: postStoreOrder,
  value: {} as Order,
});
/** auto generated by mobx-value-ts-gear-plugin */
export const mobxGetStoreOrderOrderId = request({
  request: getStoreOrderOrderId,
  value: {} as Order,
});
/** auto generated by mobx-value-ts-gear-plugin */
export const mobxDeleteStoreOrderOrderId = request({
  request: deleteStoreOrderOrderId,
  value: {} as any,
});
/** auto generated by mobx-value-ts-gear-plugin */
export const mobxPostUser = request({
  request: postUser,
  value: {} as User,
});
/** auto generated by mobx-value-ts-gear-plugin */
export const mobxPostUserCreateWithList = request({
  request: postUserCreateWithList,
  value: {} as User,
});
/** auto generated by mobx-value-ts-gear-plugin */
export const mobxGetUserLogin = request({
  request: getUserLogin,
  value: \\"\\" as string,
});
/** auto generated by mobx-value-ts-gear-plugin */
export const mobxGetUserLogout = request({
  request: getUserLogout,
  value: {} as any,
});
/** auto generated by mobx-value-ts-gear-plugin */
export const mobxGetUserUsername = request({
  request: getUserUsername,
  value: {} as User,
});
/** auto generated by mobx-value-ts-gear-plugin */
export const mobxPutUserUsername = request({
  request: putUserUsername,
  value: {} as any,
});
/** auto generated by mobx-value-ts-gear-plugin */
export const mobxDeleteUserUsername = request({
  request: deleteUserUsername,
  value: {} as any,
});
"
`;

exports[`use beforeWriteTs hook 1`] = `
"import { request } from \\"mobx-value\\";
import {
  putPet,
  postPet,
  getPetFindByStatus,
  getPetFindByTags,
  getPetPetId,
  postPetPetId,
  deletePetPetId,
  postPetPetIdUploadImage,
  getStoreInventory,
  postStoreOrder,
  getStoreOrderOrderId,
  deleteStoreOrderOrderId,
  postUser,
  postUserCreateWithList,
  getUserLogin,
  getUserLogout,
  getUserUsername,
  putUserUsername,
  deleteUserUsername,
  Pet,
  ApiResponse,
  Order,
  User,
} from \\"src/service/petv3\\";

/** auto generated by mobx-value-ts-gear-plugin */
export const requestPutPet = request({
  request: putPet,
  value: {} as Pet,
});
/** auto generated by mobx-value-ts-gear-plugin */
export const requestPostPet = request({
  request: postPet,
  value: {} as Pet,
});
/** auto generated by mobx-value-ts-gear-plugin */
export const requestGetPetFindByStatus = request({
  request: getPetFindByStatus,
  value: {} as Pet[],
});
/** auto generated by mobx-value-ts-gear-plugin */
export const requestGetPetFindByTags = request({
  request: getPetFindByTags,
  value: {} as Pet[],
});
/** auto generated by mobx-value-ts-gear-plugin */
export const requestGetPetPetId = request({
  request: getPetPetId,
  value: {} as Pet,
});
/** auto generated by mobx-value-ts-gear-plugin */
export const requestPostPetPetId = request({
  request: postPetPetId,
  value: {} as any,
});
/** auto generated by mobx-value-ts-gear-plugin */
export const requestDeletePetPetId = request({
  request: deletePetPetId,
  value: {} as any,
});
/** auto generated by mobx-value-ts-gear-plugin */
export const requestPostPetPetIdUploadImage = request({
  request: postPetPetIdUploadImage,
  value: {} as ApiResponse,
});
/** auto generated by mobx-value-ts-gear-plugin */
export const requestGetStoreInventory = request({
  request: getStoreInventory,
  value: {} as { [propertyName: string]: number },
});
/** auto generated by mobx-value-ts-gear-plugin */
export const requestPostStoreOrder = request({
  request: postStoreOrder,
  value: {} as Order,
});
/** auto generated by mobx-value-ts-gear-plugin */
export const requestGetStoreOrderOrderId = request({
  request: getStoreOrderOrderId,
  value: {} as Order,
});
/** auto generated by mobx-value-ts-gear-plugin */
export const requestDeleteStoreOrderOrderId = request({
  request: deleteStoreOrderOrderId,
  value: {} as any,
});
/** auto generated by mobx-value-ts-gear-plugin */
export const requestPostUser = request({
  request: postUser,
  value: {} as User,
});
/** auto generated by mobx-value-ts-gear-plugin */
export const requestPostUserCreateWithList = request({
  request: postUserCreateWithList,
  value: {} as User,
});
/** auto generated by mobx-value-ts-gear-plugin */
export const requestGetUserLogin = request({
  request: getUserLogin,
  value: \\"\\" as string,
});
/** auto generated by mobx-value-ts-gear-plugin */
export const requestGetUserLogout = request({
  request: getUserLogout,
  value: {} as any,
});
/** auto generated by mobx-value-ts-gear-plugin */
export const requestGetUserUsername = request({
  request: getUserUsername,
  value: {} as User,
});
/** auto generated by mobx-value-ts-gear-plugin */
export const requestPutUserUsername = request({
  request: putUserUsername,
  value: {} as any,
});
/** auto generated by mobx-value-ts-gear-plugin */
export const requestDeleteUserUsername = request({
  request: deleteUserUsername,
  value: {} as any,
});
"
`;
